pkg/settings/settings.go

```diff
@@ -1,36 +1,126 @@
+const KPM_JSON_PATH = ".kpm/config/kpm.json"
+
+// The kpm configuration
+type KpmConf struct {
+	DefaultOciRegistry string
+	DefaultOciRepo     string
+}
+
+const DEFAULT_REGISTRY = "ghcr.io"
+const DEFAULT_REPO = "KusionStack"
+
+// This is a singleton that loads kpm settings from 'kpm.json'
+// and is only initialized on the first call by 'Init()' or 'GetSettings()'
+var kpm_settings *Settings
+var once sync.Once
+
+// DefaultKpmConf create a default configuration for kpm.
+func DefaultKpmConf() KpmConf {
+	return KpmConf{
+		DefaultOciRegistry: DEFAULT_REGISTRY,
+		DefaultOciRepo:     DEFAULT_REPO,
+	}
+}
+
 type Settings struct {
 	CredentialsFile string
+	// the default configuration for kpm.
+	Conf KpmConf
 }
 
-// GetConfigJsonPath returns config.json file path under '$HOME/.kpm/config/config.json'
-func GetConfigJsonPath() (string, error) {
-	home, err := os.UserHomeDir()
+// DefaultOciRepo return the default OCI registry 'ghcr.io'.
+func (settings *Settings) DefaultOciRegistry() string {
+	return settings.Conf.DefaultOciRegistry
+}
+
+// DefaultOciRepo return the default OCI repo 'KusionStack'.
+func (settings *Settings) DefaultOciRepo() string {
+	return settings.Conf.DefaultOciRepo
+}
+
+// GetFullJsonPath returns the full path of 'config.json' and 'kpm.json' file path under '$HOME/.kpm/config/'
+func GetFullJsonPath(jsonFileName string) (string, error) {
+	home, err := env.GetAbsPkgPath()
 	if err != nil {
 		return "", errors.InternalBug
 	}
 
-	return filepath.Join(home, CONFIG_JSON_PATH), nil
+	return filepath.Join(home, jsonFileName), nil
+}
+
+// GetSettings will return the kpm setting singleton.
+func GetSettings() (*Settings, error) {
+	var err error
+	once.Do(func() {
+		credentialsFile, err := GetFullJsonPath(CONFIG_JSON_PATH)
+		if err != nil {
+			return
+		}
+
+		conf, err := loadOrCreateDefaultKpmJson()
+		if err != nil {
+			return
+		}
+
+		kpm_settings = &Settings{
+			CredentialsFile: credentialsFile,
+			Conf:            *conf,
+		}
+	})
+
+	return kpm_settings, err
 }
 
-// Init returns default kpm settings.
-func Init() (*Settings, error) {
-	credentialsFile, err := GetConfigJsonPath()
+// loadOrCreateDefaultKpmJson will load the 'kpm.json' file from '$KCL_PKG_PATH/.kpm/config',
+// and create a default 'kpm.json' file if the file does not exist.
+func loadOrCreateDefaultKpmJson() (*KpmConf, error) {
+	kpmConfpath, err := GetFullJsonPath(KPM_JSON_PATH)
 	if err != nil {
 		return nil, err
 	}
-	return &Settings{
-		CredentialsFile: credentialsFile,
-	}, nil
+
+	defaultKpmConf := DefaultKpmConf()
+
+	b, err := ioutil.ReadFile(kpmConfpath)
+	// if the file '$KCL_PKG_PATH/.kpm/config/kpm.json' does not exist
+	if os.IsNotExist(err) {
+		// create the default kpm.json.
+		err = os.MkdirAll(filepath.Dir(kpmConfpath), 0755)
+		if err != nil {
+			return nil, err
+		}
+
+		b, err := json.Marshal(defaultKpmConf)
+		if err != nil {
+			return nil, err
+		}
+		err = ioutil.WriteFile(kpmConfpath, b, 0644)
```

Comment:
os.WriteFile

Reply:
> os.WriteFile

https://github.com/helm/helm/blob/main/internal/fileutil/fileutil.go AtomicWriteFile

Reply:
Currently, kpm does not support concurrent environments, and kpm will support concurrent environments in future. This section can be adjusted in the process of enhancing concurrency in kpm.